src/lib/components/forms/Projects.svelte

```diff
@@ -0,0 +1,162 @@
+
+  import Button from "../common/Button.svelte";
+  import Field from "../inputs/Field.svelte";
+  import Flex from "../common/Flex.svelte";
+  import Switch from "../inputs/Switch.svelte";
+  import Text from "../inputs/Text.svelte";
+  import TextArea from "../inputs/TextArea.svelte";
+
+  import { getForUser, add, remove } from "$lib/api/projects";
+  import { getCsrfToken } from "$lib/utils/api";
+  import { intersection } from "@/util/array.js";
+
+  export let documents: Document[] = [];
+  export let projects: Project[] = [];
+
+  const dispatch = createEventDispatcher();
+  const me: Writable<User> = getContext("me");
+
+  let common: Set<number>;
+
+  $: common = new Set(
+    intersection(
+      documents.map((d) => d.projects),
+      (a, b) => {
+        return a.id === b.id;
+      },
+    ).map((p: Project | number) => (typeof p === "number" ? p : p.id)),
+  );
+
+  $: console.log(projects);
+
+  onMount(async () => {
+    if ($me && projects.length === 0) {
+      projects = await getForUser($me.id);
+    }
+  });
+
+  function invalidateDocs(documents: Document[]) {
+    return Promise.all(documents.map((d) => invalidate(`document:${d.id}`)));
+  }
+
+  // typescript doesn't know what to do with svelte's events
+  async function toggle(project: Project, e) {
+    const { checked } = e.target;
+    const ids = documents.map((d) => d.id);
+    const csrf_token = getCsrfToken();
+    if (checked) {
+      await add(project.id, ids, csrf_token);
+    } else {
+      await remove(project.id, ids, csrf_token);
+    }
+    await invalidateDocs(documents);
+  }
+
+  function onSubmit({ formElement, formData, action, cancel, submitter }) {
+    submitter.disabled = true;
+    return async ({ result, update }) => {
+      if (result.type === "success") {
+        projects = [...projects, result.data.project];
+        update(result);
+      }
+      submitter.disabled = false;
+    };
+  }
+</script>
+
+<div class="container">
+  <form method="post" action="/documents/projects/" use:enhance={onSubmit}>
+    <h2>{$_("projects.create")}</h2>
+    <Field title={$_("projects.fields.title")} required inline>
+      <Text name="title" placeholder={$_("projects.fields.title")} />
+    </Field>
+
+    <details>
+      <summary>{$_("common.more")} &hellip;</summary>
+      <Flex direction="column">
+        <Field title={$_("projects.fields.description")}>
+          <TextArea name="description" />
+        </Field>
+        <Field title={$_("projects.fields.private")} inline>
+          <Switch name="private" />
+        </Field>
+        <Field title={$_("projects.fields.pinned")} inline>
+          <Switch name="pinned" />
+        </Field>
+      </Flex>
+    </details>
+    <Flex>
+      <Button mode="primary" type="submit">
+        {$_("common.add")} +
+      </Button>
+    </Flex>
+  </form>
+
+  {#if projects.length}
+    <hr class="divider" />
+    <Flex direction="column" class="projects">
+      {#each projects as project}
```

Comment:
If possible, we should make sure that pinned projects are at the top of the list in alphabetic order.